### PR TITLE
Model Evaluation, Diagnostics and Registering

### LINKED ISSUE #911: Evaluation and Model Diagnostic Functions
Hello!

Related to
- #901 
- #891 

It could be nice to have some standardised model evaluation and diagnostic functions added to pymc-marketing. Ideally they'd be formulated in a way that makes them easy to log in MLFLow later on.

Standard model metrics could include:
- Bayesian R2
- MAPE, RMSE, MAE
- Normalised RMSE and MAE (to allow comparisons across datasets and methodologies - in particular with Robyn models, who use NRMSE as one of their 2 key metrics)
- etc.

Diagnostic metrics could include:
- Step size, divergences
- LOOCV metrics e.g.
```    # Calculate elemwise log_likelihood of model given posteriors
    pm.compute_log_likelihood(
        mmm.idata, model=mmm.model, progressbar=False
    )  # Used for LOOCV
    model_loo = az.loo(mmm.idata)

    # Log LOOCV metrics
    loocv_metrics = {
        "loocv_elpd_loo": model_loo.elpd_loo,  # expected log pointwise predictive density
        "loocv_se": model_loo.se,  # standard error of elpd
        "loocv_p_loo": model_loo.p_loo,  # effective number of parameters
    }```
    
    
    I will open a draft PR on this ☺️

### LINKED ISSUE #973: Add Support for Model Registering to MLFlow Module
Hello! 

I was looking through the MLFlow module and I thought it might be nice to add a wrapper to allow people to actually register their MMM in MLFlow. Having access to the Registry would unlock a lot of functionality for MMM x MLFlow users. It could be good for faster deployment / serving of MMMs, better model lifecycle management and could help with refreshes.

[MLFlow Registry](https://mlflow.org/docs/latest/model-registry.html)

I already have some code written (and tested locally) for this, which I will add in [this PR](https://github.com/pymc-labs/pymc-marketing/pull/912)

### PR DESCRIPTION
**Pull request by louismagowan**
## Description

It could be nice to have some standardised model evaluation and diagnostic functions added to pymc-marketing. Ideally they'd be formulated in a way that makes them easy to log in MLFLow later on. 
It would also be cool to build on top of the MLFlow module, to create a custom `mlflow.pyfunc.PythonModel` class to allow users to be able to register their models in the MLFlow registry. This would allow people to serve and maintain their MMMs more easily, and could help with MMM refreshes too.

Standard model metrics could include:
- Bayesian R2
- MAPE, RMSE, MAE
- Normalised RMSE and MAE (to allow comparisons across datasets and methodologies - in particular with Robyn models, who use NRMSE as one of their 2 key metrics)
etc.

Diagnostic metrics could include:
- Step size, divergences
- LOOCV metrics e.g.

Model Registry Code:
- A wrapper for an MMM model to make it conform to the MLFlow api, enabling registering and easier deployment

I'll open this as a Draft for now - since I'll need advice on how where best to put this code, as well as overall design etc.

## Related Issue

- [x] Closes #911 
- [x] Closes #973 
- [x] Related to #891 #901 

## Checklist

- [] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--912.org.readthedocs.build/en/912/

### CODE REVIEW

File: pymc_marketing/mmm/evaluation.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+def nrmse(y_true, y_pred):
+    """
+    Normalized Root Mean Square Error. Normalization allows for comparison across
+    different data sets and methodologies. e.g. NRMSE is one of the key metrics used
+    in Robyn MMMs.
+    Args:
+        y_true ([np.array]): test samples
+        y_pred ([np.array]): predicted samples
+    Returns:
+        [float]: normalized root mean square error
```
Comment (juanitorduz) [inline pymc_marketing/mmm/evaluation.py:39]: Can you please change the docstring format, see https://numpydoc.readthedocs.io/en/latest/format.html#parameters :)

Comment (juanitorduz) [reply pymc_marketing/mmm/evaluation.py:39]: See for example https://github.com/pymc-labs/pymc-marketing/blob/39f23367efe6fc512f3d0558cb97712d68d144f1/pymc_marketing/mmm/delayed_saturated_mmm.py#L227-L256

File: pymc_marketing/mmm/evaluation.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    metrics = {
+        f"{prefix}r_squared": np.nan,
+        f"{prefix}rmse": np.nan,
+        f"{prefix}nrmse": np.nan,
+        f"{prefix}mae": np.nan,
+        f"{prefix}nmae": np.nan,
+        f"{prefix}mape": np.nan,
+    }
+
+    try:
```
Comment (wd60622) [inline pymc_marketing/mmm/evaluation.py:181]: What is a reason these would error? If one fails, would they all fail?

File: pymc_marketing/mmm/evaluation.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    else:
+        prefix += "_"
+
+    metrics = {
+        f"{prefix}r_squared": np.nan,
+        f"{prefix}rmse": np.nan,
+        f"{prefix}nrmse": np.nan,
+        f"{prefix}mae": np.nan,
+        f"{prefix}nmae": np.nan,
+        f"{prefix}mape": np.nan,
```
Comment (wd60622) [inline pymc_marketing/mmm/evaluation.py:178]: How about if user can select the ones they would like?

**Comment by louismagowan** [comment]
Thanks for the feedback guys! Super helpful 😁 

I'm heading on holidays tomorrow so I might take a little while to get around to this - but keen to get stuck into it when I'm back

File: pymc_marketing/mmm/evaluation.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+import arviz as az
+import matplotlib.pyplot as plt
+import numpy as np
+import pymc as pm
+from loguru import logger
+from sklearn.metrics import (
+    mean_absolute_error,
+    mean_absolute_percentage_error,
+    root_mean_squared_error,
```
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py:22]: They don't support xarray directly - but I think something could be worked out using something like
```
hdi = az.hdi(my_posterior_samples)
lower_hdi = hdi.sel(hdi="lower").mean().y
higher_hdi = hdi.sel(hdi="higher").mean().y

lower_error = mean_absolute_error(y_true, lower_hdi)
upper_error = mean_absolute_error(y_true, upper_hdi)
```

I'm having difficulty using [`mmm.sample_posterior_predictive()`](https://github.com/pymc-labs/pymc-marketing/blob/b478d9d7beb6f980304d28f69085db77855173d3/pymc_marketing/mmm/delayed_saturated_mmm.py#L1844) to extend my idata with scaled values though. It seems like when I call that function, my idata remains unchanged

^Not sure if it's me doing something wrong or a bug though.
e.g. When I run

```
mmm.sample_posterior_predictive(X, extend_idata=True, var_names=["y"], original_scale=True)
```
I don't see those predictions anywhere in my `mmm.idata` afterwards 🤔

File: pymc_marketing/mmm/evaluation.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+    # Display LOOCV metrics (including the Pareto k values which aren't logged in MLFlow)
+    logger.info(f"LOOCV metrics: {model_loo}")
+    # Combine diagnostics and LOOCV metrics
+    model_diagnostics.update(loocv_metrics)
+
+    return model_diagnostics, model_loo
+
+
+def plot_hdi_forest(mmm: MMM, var_names: list) -> plt.Figure:
```
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py:122]: Do you have any overall design tips here @wd60622 ? ☺️ 

- Should keep this code in a `evaluate.py` module?
- Should I move it to the MLFlow module? <- Personally I think it makes more sense to keep those things separate, but happy to discuss
- Should I build an evaluation class which just inherits `ModelBuilder`?

File: pymc_marketing/mmm/evaluation.py (hunk 2 above)
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py:193]: Ah this was holdover code from an earlier version I made - I'll delete!

File: pymc_marketing/mmm/evaluation.py (hunk 1 above)
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py]: Ok fixed this part at least I think ☺️

File: pymc_marketing/mmm/evaluation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wd60622) [reply pymc_marketing/mmm/evaluation.py:23]: Can you use the extend_idata parameter?

File: pymc_marketing/mmm/evaluation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wd60622) [reply pymc_marketing/mmm/evaluation.py:131]: Think we can keep it separate then inport only the required items in mlflow module

File: pymc_marketing/mmm/evaluation.py (hunk 3 above)
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py]: Yep, good feedback - I changed this! ☺️

File: pymc_marketing/mmm/evaluation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py:23]: Yes but the `original_scale=True` argument doesn't seem to be carried through

File: pymc_marketing/mmm/evaluation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py:131]: And building `class ModelEvaluator(ModelBuilder)` is the way to go too you think? ☺️

**Comment by louismagowan** [comment]
FYI I'm still working on this and will address more of the existing feedback - just need some more time :)

File: pymc_marketing/mlflow.py
Comment (wd60622) [inline pymc_marketing/mlflow.py:1]: Think you are up for a `load_model` function? 😄 

https://github.com/mlflow/mlflow/blob/e5d1280a467f0b51692ff3edf29e191fc52f36ef/mlflow/sklearn/__init__.py#L597

File: pymc_marketing/mmm/evaluation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wd60622) [reply pymc_marketing/mmm/evaluation.py:24]: Interesting. Thanks for raising the bug. Let's address that separately

File: pymc_marketing/mmm/evaluation.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+import numpy as np
+import numpy.typing as npt
+import pymc as pm
+from sklearn.metrics import (
+    mean_absolute_error,
+    mean_absolute_percentage_error,
+    root_mean_squared_error,
+)
+
+from pymc_marketing.mmm.delayed_saturated_mmm import MMM
```
Comment (wd60622) [inline pymc_marketing/mmm/evaluation.py:27]:
```suggestion
from pymc_marketing.mmm.mmm import MMM
```

**Review by wd60622 — COMMENTED** [review]
Looking good. Various comment

**Comment by louismagowan** [comment]
Thanks again for the tips and review! Apologies for being a bit slow but work has been very busy since I got back from holidays

File: pymc_marketing/mmm/evaluation.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+        # Log divergences
+        divergences = mmm.idata["sample_stats"]["diverging"].sum().item()
+        step_size = mmm.idata["sample_stats"]["step_size"].mean().item()
+        model_diagnostics = {
+            "model_diagnostic_divergences": divergences,
+            "model_diagnostic_divergences_pct": divergences
+            / mmm.idata["sample_stats"]["diverging"].size,
+            "model_diagnostic_step_size": step_size,
+        }
```
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py:171]: Yep, I deleted from evaluate.py and just kept in mlflow.py ☺️

File: pymc_marketing/mmm/evaluation.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            mmm.idata, model=mmm.model, progressbar=False
+        )  # Used for LOOCV
+        model_loo = az.loo(mmm.idata)
+
+        # Log LOOCV metrics
+        loocv_metrics = {
+            "loocv_elpd_loo": model_loo.elpd_loo,  # expected log pointwise predictive density
+            "loocv_se": model_loo.se,  # standard error of elpd
+            "loocv_p_loo": model_loo.p_loo,  # effective number of parameters
+        }
```
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py]: Hmm yeah I was a bit stuck on this one. I moved this into mlflow.py, but I feel like that module now contains behaviour that would be useful outside the context of mlflow logging. Or perhaps, I'm adding code which to me isn't strictly "logging"

Maybe need some advice on that one 🙏

File: pymc_marketing/mlflow.py [hunk 9]
```diff
@@ -412,6 +413,189 @@ def log_inference_data(
...
+        self.combined = combined
+        self.include_last_observations = include_last_observations
+        self.original_scale = original_scale
+        self.var_names = (
+            var_names if var_names is not None else ["y"]
+        )  # Initialize if not provided
+        self.sample_kwargs = sample_kwargs
+
+    def predict(
+        self, context: Any, model_input, params: dict[str, Any] | None = None
```
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mlflow.py:592]: Yes exactly!

Unfortunately I don't think you can have a whole lot of control over the params you pass. To conform with their API I believe you can only pass in a choice of prediction method - other args won't work. I tried to get around this by having the user optionally pass in sampler_kwargs at time of class instantiation (at least then the prediction methods are set somewhat flexibly)


[Docs](https://mlflow.org/docs/latest/traditional-ml/creating-custom-pyfunc/notebooks/override-predict.html)
> We can only provide a params override for the key predict_method. Anything apart from this will be ignored and a warning will be shown indicating that the unknown parameter will not be passed to the underlying model.

File: pymc_marketing/mmm/evaluation.py
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mmm/evaluation.py:1]: Ok yep, smart!

I'll work on this 😁

**Comment by juanitorduz** [comment]
@louismagowan We had a git issue 🙈 ... I suggest you **do not pul** and create another PR to continue this great contribution ❤️

**Comment by louismagowan** [comment]
Closing this and recreating in [this PR](https://github.com/pymc-labs/pymc-marketing/pull/1026) - I'll keep referring to your existing feedback here though 😄

File: pymc_marketing/mlflow.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (louismagowan) [reply pymc_marketing/mlflow.py:594]: @wd60622  hopefully my answer was satisfactory? I'll address your other outstanding feedback on the [new PR](https://github.com/pymc-labs/pymc-marketing/pull/1026) but thought it made sense to close this one here ☺️
